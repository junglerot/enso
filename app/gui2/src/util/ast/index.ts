import * as Ast from '@/generated/ast'
import { Token, Tree } from '@/generated/ast'
import { assert } from '@/util/assert'
import { parse } from '@/util/ffi'
import { LazyObject, debug } from '@/util/parserSupport'

export { Ast }

export function parseEnso(code: string): Tree {
  const blob = parse(code)
  return Tree.read(new DataView(blob.buffer), blob.byteLength - 4)
}

/** Read a single line of code
 *
 * Is meant to be a helper for tests. If the code is multilined, an exception is raised.
 */
export function parseEnsoLine(code: string): Tree {
  const block = parseEnso(code)
  assert(block.type === Tree.Type.BodyBlock)
  const statements = block.statements[Symbol.iterator]()
  const firstLine = statements.next()
  assert(!firstLine.done)
  assert(!!statements.next().done)
  assert(firstLine.value.expression != null)
  return firstLine.value.expression
}

/**
 * Read span of code reprsented by given AST node.
 *
 * The AST is assumed to be generated from `code` and not modified sice then.
 * Otherwise an unspecified fragment of `code` may be returned.
 */
export function readAstSpan(node: Tree, code: string): string {
  const leftOffsetbegin = node.whitespaceStartInCodeParsed
  const leftOffsetEnd = leftOffsetbegin + node.whitespaceLengthInCodeParsed
  const end = leftOffsetEnd + node.childrenLengthInCodeParsed
  return code.substring(leftOffsetEnd, end)
}

/**
 * Read span of code reprsented by given Token.
 *
 * The Token is assumed to be a part of AST generated from `code`.
 */
export function readTokenSpan(token: Token, code: string): string {
  const begin = token.startInCodeBuffer
  const end = begin + token.lengthInCodeBuffer
  return code.substring(begin, end)
}

/**
 * Read direct AST children.
 */
export function childrenAstNodes(obj: LazyObject): Tree[] {
  const children: Tree[] = []
  const visitor = (obj: LazyObject) => {
    if (Tree.isInstance(obj)) {
      children.push(obj)
    } else {
      obj.visitChildren(visitor)
    }
  }
  obj.visitChildren(visitor)
  return children
}

/** Returns all AST nodes from `root` tree containing given char, starting from the most nested. */
export function astContainingChar(charIndex: number, root: Tree): Tree[] {
  return treePath(root, (node) => {
    const begin = node.whitespaceStartInCodeParsed + node.whitespaceLengthInCodeParsed
    const end = begin + node.childrenLengthInCodeParsed
    return charIndex >= begin && charIndex < end
  }).reverse()
}

/** Given a predicate, return a path from the root down the tree containing the
 *  first node at each level found to satisfy the predicate. */
function treePath(obj: LazyObject, pred: (node: Tree) => boolean): Tree[] {
  const path: Tree[] = []
  const visitor = (obj: LazyObject) => {
    if (Tree.isInstance(obj)) {
      const isMatch = pred(obj)
      if (isMatch) path.push(obj)
      return obj.visitChildren(visitor) || isMatch
    } else {
      return obj.visitChildren(visitor)
    }
  }
  obj.visitChildren(visitor)
  return path
}

if (import.meta.vitest) {
  const { test, expect } = import.meta.vitest

  const parseCases = [
    'foo bar\n',
    'Data.read\n2 + 2',
    'Data.read File\n2 + 3',
    'Data.read "File"\n2 + 3',
    'foo bar=baz',
    '2\n + 3\n + 4',
  ]

  test.each(parseCases)("Parsing '%s'", (code) => {
    expect(debug(parseEnso(code))).toMatchSnapshot()
  })

  test.each(parseCases)("AST spans of '%s' are valid", (input) => {
    const tree = parseEnso(input)
    const endPos = validateSpans(tree)
    expect(endPos).toStrictEqual(input.length)
  })

  test("Reading AST node's code", () => {
    const code = 'Data.read File\n2 + 3'
    const ast = parseEnso(code)
    expect(readAstSpan(ast, code)).toStrictEqual(code)
    assert(ast.type === Tree.Type.BodyBlock)
    const statements = Array.from(ast.statements)

    assert(statements[0]?.expression != null)
    expect(readAstSpan(statements[0].expression, code)).toStrictEqual('Data.read File')
    assert(statements[0].expression.type === Tree.Type.App)
    expect(readAstSpan(statements[0].expression.func, code)).toStrictEqual('Data.read')
    expect(readAstSpan(statements[0].expression.arg, code)).toStrictEqual('File')

    assert(statements[1]?.expression != null)
    expect(readAstSpan(statements[1].expression, code)).toStrictEqual('2 + 3')
    assert(statements[1].expression.type === Tree.Type.OprApp)
    assert(statements[1].expression.lhs != null)
    assert(statements[1].expression.rhs != null)
    assert(statements[1].expression.opr.ok)
    expect(readAstSpan(statements[1].expression.lhs, code)).toStrictEqual('2')
    expect(readTokenSpan(statements[1].expression.opr.value, code)).toStrictEqual('+')
    expect(readAstSpan(statements[1].expression.rhs, code)).toStrictEqual('3')
  })

  test.each([
    [
      '2 + a',
      [
        { type: Tree.Type.Number, repr: '2' },
        { type: Tree.Type.Ident, repr: 'a' },
      ],
    ],
    [
      'a.b',
      [
        { type: Tree.Type.Ident, repr: 'a' },
        { type: Tree.Type.Ident, repr: 'b' },
      ],
    ],
    [
      'Data.read foo',
      [
        { type: Tree.Type.OprApp, repr: 'Data.read' },
        { type: Tree.Type.Ident, repr: 'foo' },
      ],
    ],
    ['(2 + a)', [{ type: Tree.Type.OprApp, repr: '2 + a' }]],
    [
      'Data.read\n  foo\n  bar',
      [
        { type: Tree.Type.OprApp, repr: 'Data.read' },
        { type: Tree.Type.Ident, repr: 'foo' },
        { type: Tree.Type.Ident, repr: 'bar' },
      ],
    ],
    [
      'Data.read file=foo',
      [
        { type: Tree.Type.OprApp, repr: 'Data.read' },
        { type: Tree.Type.Ident, repr: 'foo' },
      ],
    ],
  ])("Reading children of '%s'", (code, expected) => {
    const ast = parseEnsoLine(code)
    const children = Array.from(childrenAstNodes(ast))
    const childrenWithExpected = children.map((child, i) => {
      return { child, expected: expected[i] }
    })
    for (const { child, expected } of childrenWithExpected) {
      expect(child.type).toBe(expected?.type)
      expect(readAstSpan(child, code)).toBe(expected?.repr)
    }
  })

  test.each([
    [
      '2 + a',
      0,
      [
        { type: Tree.Type.Number, repr: '2' },
        { type: Tree.Type.OprApp, repr: '2 + a' },
        { type: Tree.Type.BodyBlock, repr: '2 + a' },
      ],
    ],
    [
      'Data.read foo',
      5,
      [
        { type: Tree.Type.Ident, repr: 'read' },
        { type: Tree.Type.OprApp, repr: 'Data.read' },
        { type: Tree.Type.App, repr: 'Data.read foo' },
        { type: Tree.Type.BodyBlock, repr: 'Data.read foo' },
      ],
    ],
    [
      'Data.read foo',
      4,
      [
        { type: Tree.Type.OprApp, repr: 'Data.read' },
        { type: Tree.Type.App, repr: 'Data.read foo' },
        { type: Tree.Type.BodyBlock, repr: 'Data.read foo' },
      ],
    ],
    [
      'Data.read foo',
      9,
      [
        { type: Tree.Type.App, repr: 'Data.read foo' },
        { type: Tree.Type.BodyBlock, repr: 'Data.read foo' },
      ],
    ],
    [
      'Data.',
      4,
      [
        { type: Tree.Type.OprApp, repr: 'Data.' },
        { type: Tree.Type.OprSectionBoundary, repr: 'Data.' },
        { type: Tree.Type.BodyBlock, repr: 'Data.' },
      ],
    ],
  ])("Reading AST from code '%s' and position %i", (code, position, expected) => {
    const ast = parseEnso(code)
    const astAtPosition = astContainingChar(position, ast)
    const resultWithExpected = astAtPosition.map((ast, i) => {
      return { ast, expected: expected[i] }
    })
    for (const { ast, expected } of resultWithExpected) {
      expect(ast.type).toBe(expected?.type)
      expect(readAstSpan(ast, code)).toBe(expected?.repr)
    }
  })
}

function validateSpans(obj: LazyObject, initialPos?: number): number {
  const state = { pos: initialPos ?? 0 }
  const visitor = (value: LazyObject) => {
    if (
      Token.isInstance(value) &&
      !(value.whitespaceLengthInCodeBuffer + value.lengthInCodeBuffer === 0)
    ) {
      assert(value.whitespaceStartInCodeBuffer === state.pos)
      state.pos += value.whitespaceLengthInCodeBuffer
      assert(value.startInCodeBuffer === state.pos)
      state.pos += value.lengthInCodeBuffer
    } else if (Tree.isInstance(value)) {
      assert(value.whitespaceStartInCodeParsed === state.pos)
      state.pos += value.whitespaceLengthInCodeParsed
      const end = state.pos + value.childrenLengthInCodeParsed
      value.visitChildren(visitor)
      assert(state.pos === end)
    } else {
      value.visitChildren(visitor)
    }
  }
  visitor(obj)
  return state.pos
}
