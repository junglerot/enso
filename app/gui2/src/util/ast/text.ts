const mapping: Record<string, string> = {
  '\b': '\\b',
  '\f': '\\f',
  '\n': '\\n',
  '\r': '\\r',
  '\t': '\\t',
  '\v': '\\v',
  '\\': '\\\\',
  '"': '\\"',
  "'": "\\'",
  '`': '``',
}

/** Escape a string so it can be safely spliced into an interpolated (`''`) Enso string.
 * NOT USABLE to insert into raw strings. Does not include quotes. */
export function escape(string: string) {
  return string.replace(/[\0\b\f\n\r\t\v\\"'`]/g, (match) => mapping[match]!)
}
