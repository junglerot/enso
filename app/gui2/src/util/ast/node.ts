import type { Node } from '@/stores/graph'
import { Ast } from '@/util/ast'
import { Vec2 } from '@/util/data/vec2'

export function nodeFromAst(ast: Ast.Ast): Node {
  if (ast instanceof Ast.Assignment) {
    return {
      outerExprId: ast.id,
      pattern: ast.pattern ?? undefined,
      rootSpan: ast.expression ?? ast,
      position: Vec2.Zero,
      vis: undefined,
    }
  } else {
    return {
      outerExprId: ast.id,
      pattern: undefined,
      rootSpan: ast,
      position: Vec2.Zero,
      vis: undefined,
    }
  }
}
