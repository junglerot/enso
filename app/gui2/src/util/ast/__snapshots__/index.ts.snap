// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Parsing ' foo bar
' 1`] = `
{
  "childrenLengthInCodeParsed": 8,
  "statements": [
    {
      "expression": {
        "arg": {
          "childrenLengthInCodeParsed": 3,
          "token": {
            "isFree": false,
            "isOperatorLexically": false,
            "isTypeOrConstructor": false,
            "lengthInCodeBuffer": 3,
            "liftLevel": 0,
            "startInCodeBuffer": 5,
            "type": "Ident",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 5,
          },
          "type": "Ident",
          "whitespaceLengthInCodeParsed": 1,
          "whitespaceStartInCodeParsed": 4,
        },
        "childrenLengthInCodeParsed": 7,
        "func": {
          "childrenLengthInCodeParsed": 3,
          "token": {
            "isFree": false,
            "isOperatorLexically": false,
            "isTypeOrConstructor": false,
            "lengthInCodeBuffer": 3,
            "liftLevel": 0,
            "startInCodeBuffer": 1,
            "type": "Ident",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 1,
          },
          "type": "Ident",
          "whitespaceLengthInCodeParsed": 0,
          "whitespaceStartInCodeParsed": 1,
        },
        "type": "App",
        "whitespaceLengthInCodeParsed": 0,
        "whitespaceStartInCodeParsed": 1,
      },
      "newline": {
        "lengthInCodeBuffer": 0,
        "startInCodeBuffer": 0,
        "type": "Newline",
        "whitespaceLengthInCodeBuffer": 0,
        "whitespaceStartInCodeBuffer": 0,
      },
    },
    {
      "expression": undefined,
      "newline": {
        "lengthInCodeBuffer": 1,
        "startInCodeBuffer": 8,
        "type": "Newline",
        "whitespaceLengthInCodeBuffer": 0,
        "whitespaceStartInCodeBuffer": 8,
      },
    },
  ],
  "type": "BodyBlock",
  "whitespaceLengthInCodeParsed": 1,
  "whitespaceStartInCodeParsed": 0,
}
`;

exports[`Parsing '2
 + 3
 + 4' 1`] = `
{
  "childrenLengthInCodeParsed": 11,
  "statements": [
    {
      "expression": {
        "childrenLengthInCodeParsed": 11,
        "excess": [],
        "expressions": [
          {
            "expression": {
              "expression": {
                "base": undefined,
                "childrenLengthInCodeParsed": 1,
                "fractionalDigits": undefined,
                "integer": {
                  "base": undefined,
                  "lengthInCodeBuffer": 1,
                  "startInCodeBuffer": 5,
                  "type": "Digits",
                  "whitespaceLengthInCodeBuffer": 0,
                  "whitespaceStartInCodeBuffer": 5,
                },
                "type": "Number",
                "whitespaceLengthInCodeParsed": 1,
                "whitespaceStartInCodeParsed": 4,
              },
              "operator": {
                "ok": true,
                "value": {
                  "lengthInCodeBuffer": 1,
                  "startInCodeBuffer": 3,
                  "type": "Operator",
                  "whitespaceLengthInCodeBuffer": 1,
                  "whitespaceStartInCodeBuffer": 2,
                },
              },
            },
            "newline": {
              "lengthInCodeBuffer": 1,
              "startInCodeBuffer": 1,
              "type": "Newline",
              "whitespaceLengthInCodeBuffer": 0,
              "whitespaceStartInCodeBuffer": 1,
            },
          },
          {
            "expression": {
              "expression": {
                "base": undefined,
                "childrenLengthInCodeParsed": 1,
                "fractionalDigits": undefined,
                "integer": {
                  "base": undefined,
                  "lengthInCodeBuffer": 1,
                  "startInCodeBuffer": 10,
                  "type": "Digits",
                  "whitespaceLengthInCodeBuffer": 0,
                  "whitespaceStartInCodeBuffer": 10,
                },
                "type": "Number",
                "whitespaceLengthInCodeParsed": 1,
                "whitespaceStartInCodeParsed": 9,
              },
              "operator": {
                "ok": true,
                "value": {
                  "lengthInCodeBuffer": 1,
                  "startInCodeBuffer": 8,
                  "type": "Operator",
                  "whitespaceLengthInCodeBuffer": 1,
                  "whitespaceStartInCodeBuffer": 7,
                },
              },
            },
            "newline": {
              "lengthInCodeBuffer": 1,
              "startInCodeBuffer": 6,
              "type": "Newline",
              "whitespaceLengthInCodeBuffer": 0,
              "whitespaceStartInCodeBuffer": 6,
            },
          },
        ],
        "lhs": {
          "base": undefined,
          "childrenLengthInCodeParsed": 1,
          "fractionalDigits": undefined,
          "integer": {
            "base": undefined,
            "lengthInCodeBuffer": 1,
            "startInCodeBuffer": 0,
            "type": "Digits",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 0,
          },
          "type": "Number",
          "whitespaceLengthInCodeParsed": 0,
          "whitespaceStartInCodeParsed": 0,
        },
        "type": "OperatorBlockApplication",
        "whitespaceLengthInCodeParsed": 0,
        "whitespaceStartInCodeParsed": 0,
      },
      "newline": {
        "lengthInCodeBuffer": 0,
        "startInCodeBuffer": 0,
        "type": "Newline",
        "whitespaceLengthInCodeBuffer": 0,
        "whitespaceStartInCodeBuffer": 0,
      },
    },
  ],
  "type": "BodyBlock",
  "whitespaceLengthInCodeParsed": 0,
  "whitespaceStartInCodeParsed": 0,
}
`;

exports[`Parsing 'Data.read
2 + 2' 1`] = `
{
  "childrenLengthInCodeParsed": 15,
  "statements": [
    {
      "expression": {
        "childrenLengthInCodeParsed": 9,
        "lhs": {
          "childrenLengthInCodeParsed": 4,
          "token": {
            "isFree": false,
            "isOperatorLexically": false,
            "isTypeOrConstructor": true,
            "lengthInCodeBuffer": 4,
            "liftLevel": 0,
            "startInCodeBuffer": 0,
            "type": "Ident",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 0,
          },
          "type": "Ident",
          "whitespaceLengthInCodeParsed": 0,
          "whitespaceStartInCodeParsed": 0,
        },
        "opr": {
          "ok": true,
          "value": {
            "lengthInCodeBuffer": 1,
            "startInCodeBuffer": 4,
            "type": "Operator",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 4,
          },
        },
        "rhs": {
          "childrenLengthInCodeParsed": 4,
          "token": {
            "isFree": false,
            "isOperatorLexically": false,
            "isTypeOrConstructor": false,
            "lengthInCodeBuffer": 4,
            "liftLevel": 0,
            "startInCodeBuffer": 5,
            "type": "Ident",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 5,
          },
          "type": "Ident",
          "whitespaceLengthInCodeParsed": 0,
          "whitespaceStartInCodeParsed": 5,
        },
        "type": "OprApp",
        "whitespaceLengthInCodeParsed": 0,
        "whitespaceStartInCodeParsed": 0,
      },
      "newline": {
        "lengthInCodeBuffer": 0,
        "startInCodeBuffer": 0,
        "type": "Newline",
        "whitespaceLengthInCodeBuffer": 0,
        "whitespaceStartInCodeBuffer": 0,
      },
    },
    {
      "expression": {
        "childrenLengthInCodeParsed": 5,
        "lhs": {
          "base": undefined,
          "childrenLengthInCodeParsed": 1,
          "fractionalDigits": undefined,
          "integer": {
            "base": undefined,
            "lengthInCodeBuffer": 1,
            "startInCodeBuffer": 10,
            "type": "Digits",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 10,
          },
          "type": "Number",
          "whitespaceLengthInCodeParsed": 0,
          "whitespaceStartInCodeParsed": 10,
        },
        "opr": {
          "ok": true,
          "value": {
            "lengthInCodeBuffer": 1,
            "startInCodeBuffer": 12,
            "type": "Operator",
            "whitespaceLengthInCodeBuffer": 1,
            "whitespaceStartInCodeBuffer": 11,
          },
        },
        "rhs": {
          "base": undefined,
          "childrenLengthInCodeParsed": 1,
          "fractionalDigits": undefined,
          "integer": {
            "base": undefined,
            "lengthInCodeBuffer": 1,
            "startInCodeBuffer": 14,
            "type": "Digits",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 14,
          },
          "type": "Number",
          "whitespaceLengthInCodeParsed": 1,
          "whitespaceStartInCodeParsed": 13,
        },
        "type": "OprApp",
        "whitespaceLengthInCodeParsed": 0,
        "whitespaceStartInCodeParsed": 10,
      },
      "newline": {
        "lengthInCodeBuffer": 1,
        "startInCodeBuffer": 9,
        "type": "Newline",
        "whitespaceLengthInCodeBuffer": 0,
        "whitespaceStartInCodeBuffer": 9,
      },
    },
  ],
  "type": "BodyBlock",
  "whitespaceLengthInCodeParsed": 0,
  "whitespaceStartInCodeParsed": 0,
}
`;

exports[`Parsing 'Data.read "File"
2 + 3' 1`] = `
{
  "childrenLengthInCodeParsed": 22,
  "statements": [
    {
      "expression": {
        "arg": {
          "childrenLengthInCodeParsed": 6,
          "close": {
            "lengthInCodeBuffer": 1,
            "startInCodeBuffer": 15,
            "type": "TextEnd",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 15,
          },
          "elements": [
            {
              "text": {
                "lengthInCodeBuffer": 4,
                "startInCodeBuffer": 11,
                "type": "TextSection",
                "whitespaceLengthInCodeBuffer": 0,
                "whitespaceStartInCodeBuffer": 11,
              },
              "type": "Section",
            },
          ],
          "newline": undefined,
          "open": {
            "lengthInCodeBuffer": 1,
            "startInCodeBuffer": 10,
            "type": "TextStart",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 10,
          },
          "type": "TextLiteral",
          "whitespaceLengthInCodeParsed": 1,
          "whitespaceStartInCodeParsed": 9,
        },
        "childrenLengthInCodeParsed": 16,
        "func": {
          "childrenLengthInCodeParsed": 9,
          "lhs": {
            "childrenLengthInCodeParsed": 4,
            "token": {
              "isFree": false,
              "isOperatorLexically": false,
              "isTypeOrConstructor": true,
              "lengthInCodeBuffer": 4,
              "liftLevel": 0,
              "startInCodeBuffer": 0,
              "type": "Ident",
              "whitespaceLengthInCodeBuffer": 0,
              "whitespaceStartInCodeBuffer": 0,
            },
            "type": "Ident",
            "whitespaceLengthInCodeParsed": 0,
            "whitespaceStartInCodeParsed": 0,
          },
          "opr": {
            "ok": true,
            "value": {
              "lengthInCodeBuffer": 1,
              "startInCodeBuffer": 4,
              "type": "Operator",
              "whitespaceLengthInCodeBuffer": 0,
              "whitespaceStartInCodeBuffer": 4,
            },
          },
          "rhs": {
            "childrenLengthInCodeParsed": 4,
            "token": {
              "isFree": false,
              "isOperatorLexically": false,
              "isTypeOrConstructor": false,
              "lengthInCodeBuffer": 4,
              "liftLevel": 0,
              "startInCodeBuffer": 5,
              "type": "Ident",
              "whitespaceLengthInCodeBuffer": 0,
              "whitespaceStartInCodeBuffer": 5,
            },
            "type": "Ident",
            "whitespaceLengthInCodeParsed": 0,
            "whitespaceStartInCodeParsed": 5,
          },
          "type": "OprApp",
          "whitespaceLengthInCodeParsed": 0,
          "whitespaceStartInCodeParsed": 0,
        },
        "type": "App",
        "whitespaceLengthInCodeParsed": 0,
        "whitespaceStartInCodeParsed": 0,
      },
      "newline": {
        "lengthInCodeBuffer": 0,
        "startInCodeBuffer": 0,
        "type": "Newline",
        "whitespaceLengthInCodeBuffer": 0,
        "whitespaceStartInCodeBuffer": 0,
      },
    },
    {
      "expression": {
        "childrenLengthInCodeParsed": 5,
        "lhs": {
          "base": undefined,
          "childrenLengthInCodeParsed": 1,
          "fractionalDigits": undefined,
          "integer": {
            "base": undefined,
            "lengthInCodeBuffer": 1,
            "startInCodeBuffer": 17,
            "type": "Digits",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 17,
          },
          "type": "Number",
          "whitespaceLengthInCodeParsed": 0,
          "whitespaceStartInCodeParsed": 17,
        },
        "opr": {
          "ok": true,
          "value": {
            "lengthInCodeBuffer": 1,
            "startInCodeBuffer": 19,
            "type": "Operator",
            "whitespaceLengthInCodeBuffer": 1,
            "whitespaceStartInCodeBuffer": 18,
          },
        },
        "rhs": {
          "base": undefined,
          "childrenLengthInCodeParsed": 1,
          "fractionalDigits": undefined,
          "integer": {
            "base": undefined,
            "lengthInCodeBuffer": 1,
            "startInCodeBuffer": 21,
            "type": "Digits",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 21,
          },
          "type": "Number",
          "whitespaceLengthInCodeParsed": 1,
          "whitespaceStartInCodeParsed": 20,
        },
        "type": "OprApp",
        "whitespaceLengthInCodeParsed": 0,
        "whitespaceStartInCodeParsed": 17,
      },
      "newline": {
        "lengthInCodeBuffer": 1,
        "startInCodeBuffer": 16,
        "type": "Newline",
        "whitespaceLengthInCodeBuffer": 0,
        "whitespaceStartInCodeBuffer": 16,
      },
    },
  ],
  "type": "BodyBlock",
  "whitespaceLengthInCodeParsed": 0,
  "whitespaceStartInCodeParsed": 0,
}
`;

exports[`Parsing 'Data.read File
2 + 3' 1`] = `
{
  "childrenLengthInCodeParsed": 20,
  "statements": [
    {
      "expression": {
        "arg": {
          "childrenLengthInCodeParsed": 4,
          "token": {
            "isFree": false,
            "isOperatorLexically": false,
            "isTypeOrConstructor": true,
            "lengthInCodeBuffer": 4,
            "liftLevel": 0,
            "startInCodeBuffer": 10,
            "type": "Ident",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 10,
          },
          "type": "Ident",
          "whitespaceLengthInCodeParsed": 1,
          "whitespaceStartInCodeParsed": 9,
        },
        "childrenLengthInCodeParsed": 14,
        "func": {
          "childrenLengthInCodeParsed": 9,
          "lhs": {
            "childrenLengthInCodeParsed": 4,
            "token": {
              "isFree": false,
              "isOperatorLexically": false,
              "isTypeOrConstructor": true,
              "lengthInCodeBuffer": 4,
              "liftLevel": 0,
              "startInCodeBuffer": 0,
              "type": "Ident",
              "whitespaceLengthInCodeBuffer": 0,
              "whitespaceStartInCodeBuffer": 0,
            },
            "type": "Ident",
            "whitespaceLengthInCodeParsed": 0,
            "whitespaceStartInCodeParsed": 0,
          },
          "opr": {
            "ok": true,
            "value": {
              "lengthInCodeBuffer": 1,
              "startInCodeBuffer": 4,
              "type": "Operator",
              "whitespaceLengthInCodeBuffer": 0,
              "whitespaceStartInCodeBuffer": 4,
            },
          },
          "rhs": {
            "childrenLengthInCodeParsed": 4,
            "token": {
              "isFree": false,
              "isOperatorLexically": false,
              "isTypeOrConstructor": false,
              "lengthInCodeBuffer": 4,
              "liftLevel": 0,
              "startInCodeBuffer": 5,
              "type": "Ident",
              "whitespaceLengthInCodeBuffer": 0,
              "whitespaceStartInCodeBuffer": 5,
            },
            "type": "Ident",
            "whitespaceLengthInCodeParsed": 0,
            "whitespaceStartInCodeParsed": 5,
          },
          "type": "OprApp",
          "whitespaceLengthInCodeParsed": 0,
          "whitespaceStartInCodeParsed": 0,
        },
        "type": "App",
        "whitespaceLengthInCodeParsed": 0,
        "whitespaceStartInCodeParsed": 0,
      },
      "newline": {
        "lengthInCodeBuffer": 0,
        "startInCodeBuffer": 0,
        "type": "Newline",
        "whitespaceLengthInCodeBuffer": 0,
        "whitespaceStartInCodeBuffer": 0,
      },
    },
    {
      "expression": {
        "childrenLengthInCodeParsed": 5,
        "lhs": {
          "base": undefined,
          "childrenLengthInCodeParsed": 1,
          "fractionalDigits": undefined,
          "integer": {
            "base": undefined,
            "lengthInCodeBuffer": 1,
            "startInCodeBuffer": 15,
            "type": "Digits",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 15,
          },
          "type": "Number",
          "whitespaceLengthInCodeParsed": 0,
          "whitespaceStartInCodeParsed": 15,
        },
        "opr": {
          "ok": true,
          "value": {
            "lengthInCodeBuffer": 1,
            "startInCodeBuffer": 17,
            "type": "Operator",
            "whitespaceLengthInCodeBuffer": 1,
            "whitespaceStartInCodeBuffer": 16,
          },
        },
        "rhs": {
          "base": undefined,
          "childrenLengthInCodeParsed": 1,
          "fractionalDigits": undefined,
          "integer": {
            "base": undefined,
            "lengthInCodeBuffer": 1,
            "startInCodeBuffer": 19,
            "type": "Digits",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 19,
          },
          "type": "Number",
          "whitespaceLengthInCodeParsed": 1,
          "whitespaceStartInCodeParsed": 18,
        },
        "type": "OprApp",
        "whitespaceLengthInCodeParsed": 0,
        "whitespaceStartInCodeParsed": 15,
      },
      "newline": {
        "lengthInCodeBuffer": 1,
        "startInCodeBuffer": 14,
        "type": "Newline",
        "whitespaceLengthInCodeBuffer": 0,
        "whitespaceStartInCodeBuffer": 14,
      },
    },
  ],
  "type": "BodyBlock",
  "whitespaceLengthInCodeParsed": 0,
  "whitespaceStartInCodeParsed": 0,
}
`;

exports[`Parsing 'foo bar
' 1`] = `
{
  "childrenLengthInCodeParsed": 8,
  "statements": [
    {
      "expression": {
        "arg": {
          "childrenLengthInCodeParsed": 3,
          "token": {
            "isFree": false,
            "isOperatorLexically": false,
            "isTypeOrConstructor": false,
            "lengthInCodeBuffer": 3,
            "liftLevel": 0,
            "startInCodeBuffer": 4,
            "type": "Ident",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 4,
          },
          "type": "Ident",
          "whitespaceLengthInCodeParsed": 1,
          "whitespaceStartInCodeParsed": 3,
        },
        "childrenLengthInCodeParsed": 7,
        "func": {
          "childrenLengthInCodeParsed": 3,
          "token": {
            "isFree": false,
            "isOperatorLexically": false,
            "isTypeOrConstructor": false,
            "lengthInCodeBuffer": 3,
            "liftLevel": 0,
            "startInCodeBuffer": 0,
            "type": "Ident",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 0,
          },
          "type": "Ident",
          "whitespaceLengthInCodeParsed": 0,
          "whitespaceStartInCodeParsed": 0,
        },
        "type": "App",
        "whitespaceLengthInCodeParsed": 0,
        "whitespaceStartInCodeParsed": 0,
      },
      "newline": {
        "lengthInCodeBuffer": 0,
        "startInCodeBuffer": 0,
        "type": "Newline",
        "whitespaceLengthInCodeBuffer": 0,
        "whitespaceStartInCodeBuffer": 0,
      },
    },
    {
      "expression": undefined,
      "newline": {
        "lengthInCodeBuffer": 1,
        "startInCodeBuffer": 7,
        "type": "Newline",
        "whitespaceLengthInCodeBuffer": 0,
        "whitespaceStartInCodeBuffer": 7,
      },
    },
  ],
  "type": "BodyBlock",
  "whitespaceLengthInCodeParsed": 0,
  "whitespaceStartInCodeParsed": 0,
}
`;

exports[`Parsing 'foo bar=baz' 1`] = `
{
  "childrenLengthInCodeParsed": 11,
  "statements": [
    {
      "expression": {
        "arg": {
          "childrenLengthInCodeParsed": 3,
          "token": {
            "isFree": false,
            "isOperatorLexically": false,
            "isTypeOrConstructor": false,
            "lengthInCodeBuffer": 3,
            "liftLevel": 0,
            "startInCodeBuffer": 8,
            "type": "Ident",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 8,
          },
          "type": "Ident",
          "whitespaceLengthInCodeParsed": 0,
          "whitespaceStartInCodeParsed": 8,
        },
        "childrenLengthInCodeParsed": 11,
        "close": undefined,
        "equals": {
          "lengthInCodeBuffer": 1,
          "startInCodeBuffer": 7,
          "type": "Operator",
          "whitespaceLengthInCodeBuffer": 0,
          "whitespaceStartInCodeBuffer": 7,
        },
        "func": {
          "childrenLengthInCodeParsed": 3,
          "token": {
            "isFree": false,
            "isOperatorLexically": false,
            "isTypeOrConstructor": false,
            "lengthInCodeBuffer": 3,
            "liftLevel": 0,
            "startInCodeBuffer": 0,
            "type": "Ident",
            "whitespaceLengthInCodeBuffer": 0,
            "whitespaceStartInCodeBuffer": 0,
          },
          "type": "Ident",
          "whitespaceLengthInCodeParsed": 0,
          "whitespaceStartInCodeParsed": 0,
        },
        "name": {
          "isFree": false,
          "isOperatorLexically": false,
          "isTypeOrConstructor": false,
          "lengthInCodeBuffer": 3,
          "liftLevel": 0,
          "startInCodeBuffer": 4,
          "type": "Ident",
          "whitespaceLengthInCodeBuffer": 1,
          "whitespaceStartInCodeBuffer": 3,
        },
        "open": undefined,
        "type": "NamedApp",
        "whitespaceLengthInCodeParsed": 0,
        "whitespaceStartInCodeParsed": 0,
      },
      "newline": {
        "lengthInCodeBuffer": 0,
        "startInCodeBuffer": 0,
        "type": "Newline",
        "whitespaceLengthInCodeBuffer": 0,
        "whitespaceStartInCodeBuffer": 0,
      },
    },
  ],
  "type": "BodyBlock",
  "whitespaceLengthInCodeParsed": 0,
  "whitespaceStartInCodeParsed": 0,
}
`;
